File: src/timescatter.R

#Get TCR data
tcr <- read.table('../results/timescatter/volunteercounts.txt',header=TRUE,sep=',',row.names='volunteer')

#Make plots
require('ggplot2')
require('ggpubr')
require('gridExtra')

time<-data.frame(volunteer=rep(row.names(tcr),2),
                 entropy=c(tcr$B0entropy,tcr$B60entropy),
                 total=c(tcr$B0total,tcr$B60total),
                 reads=c(tcr$B0reads,tcr$B60reads),
                 PPIntfreq=c(tcr$PPIntB0freq,tcr$PPIntB60freq),
                 day=c(rep('Day0',33),rep('Day60',33)),
                 responder=rep(tcr$responder,2))

p<-ggplot(tcr,aes(x=B0total,y=B60total,shape=responder)) + geom_point() + theme_classic()
p<-p + xlab("Day 0 unique TCRs") + ylab("Day 60 unique TCRs")
p
ggsave('../results/uniqueTCRs.pdf',p)

p<-ggplot(tcr,aes(x=B0reads,y=B60reads,label=row.names(tcr))) + geom_text() + theme_classic()
p<-p + xlab("Day 0 TCRs reads") + ylab("Day 60 TCRs reads")
p
ggsave('../results/timescatter/TCRreads.pdf',p)

p<-ggboxplot(time,y="PPIntfreq",x="day", color="day")
p<-p + stat_compare_means(method = "wilcox.test", paired = TRUE, aes(label = ..p.signif..),label.x = 1.5)
p<-p+geom_line(aes(group = volunteer),alpha = 0.1) + theme(legend.position = "none") 
p<-p+ylab("HBV-specific T-cell frequency")+xlab("")
p
ggsave('../results/timescatter/PPIntfreq.pdf',p)

pte<-ggboxplot(time[time['responder']=='Early-converter',],y="total",x="day", color="day")
pte<-pte + stat_compare_means(method = "wilcox.test", paired = TRUE, aes(label = ..p.signif..),label.x = 1.5)
pte<-pte +geom_line(aes(group = volunteer),alpha = 0.1)
pte<-pte + xlab("") + ylab("Unique CDR3s") + ggtitle("Early-converter") + theme(plot.title = element_text(hjust = 0.5))
pte<-pte + theme(legend.position = "none") 

ptl<-ggboxplot(time[time['responder']=='Late-converter',],y="total",x="day", color="day")
ptl<-ptl + stat_compare_means(method = "wilcox.test", paired = TRUE, aes(label = ..p.signif..),label.x = 1.5)
ptl<-ptl +geom_line(aes(group = volunteer),alpha = 0.1)
ptl<-ptl + xlab("") + ylab("")  + ggtitle("Late-converter") + theme(plot.title = element_text(hjust = 0.5))
ptl<-ptl + theme(legend.position = "none") 

pee<-ggboxplot(time[time['responder']=='Early-converter',],y="entropy",x="day", color="day")
pee<-pee + stat_compare_means(method = "wilcox.test", paired = TRUE, aes(label = ..p.signif..),label.x = 1.5)
pee<-pee +geom_line(aes(group = volunteer),alpha = 0.1)
pee<-pee + xlab("") + ylab("Entropy") + ylim(6,8)
pee<-pee + theme(legend.position = "none") 

pel<-ggboxplot(time[time['responder']=='Late-converter',],y="entropy",x="day", color="day")
pel<-pel + stat_compare_means(method = "wilcox.test", paired = TRUE, aes(label = ..p.signif..),label.x = 1.5)
pel<-pel +geom_line(aes(group = volunteer),alpha = 0.1)
pel<-pel + xlab("") + ylab("") +ylim(6,8)
pel<-pel + theme(legend.position = "none") 

etplot<-grid.arrange(pte, ptl, pee, pel, nrow = 2, ncol= 2)
ggsave('../results/entropy-total.pdf',etplot)

wilcox.test(tcr$B0entropy,tcr$B60entropy,paired=TRUE)
wilcox.test(tcr$PPIntB0freq,tcr$PPIntB60freq,paired=TRUE)
wilcox.test(tcr[tcr['responder']=='Late-converter',]$B0entropy,tcr[tcr['responder']=='Late-converter',]$B60entropy,paired=TRUE)
wilcox.test(tcr[tcr['responder']=='Late-converter',]$B0total,tcr[tcr['responder']=='Late-converter',]$B60total,paired=TRUE)
wilcox.test(tcr[tcr['responder']=='Late-converter',]$PPIntB0freq,tcr[tcr['responder']=='Late-converter',]$PPIntB60freq,paired=TRUE)


wilcox.test(tcr[tcr['responder']=='Early-converter',]$B0entropy,tcr[tcr['responder']=='Early-converter',]$B60entropy,paired=TRUE)
wilcox.test(tcr[tcr['responder']=='Early-converter',]$B0total,tcr[tcr['responder']=='Early-converter',]$B60total,paired=TRUE)
wilcox.test(tcr[tcr['responder']=='Early-converter',]$PPIntB0freq,tcr[tcr['responder']=='Early-converter',]$PPIntB60freq,paired=TRUE)


# 


# Supplemental figure

supp<-data.frame(vaccinee = rep(tcr$X,3))
supp$TCRcounts <- c(tcr$B0B60overlap, (tcr$B0total - tcr$B0B60overlap), ( tcr$B60total - tcr$B0B60overlap))
supp$label <- c(rep("Overlap",33),rep("Day0 Unique",33),rep("Day60 Unique",33))

s.unique<-ggplot(supp,aes(x=vaccinee,y=TCRcounts,fill=label)) + geom_col() 
s.unique<-s.unique+ ylab("Unique TCRs") + theme_classic() + theme(legend.title = element_blank(), axis.text.x=element_blank(), axis.ticks.x=element_blank()) + xlab("Vaccinees")
s.unique
ggsave('../results/s_unique.pdf',s.unique)

File: src/timescatter.py
#!/usr/local/bin/python3
# Script to generate scattr plots to compare day 0 and day 60 TCR repertoires
# Colors overlay identified peptide / pepetide pool specific TCRs

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import re

from tcrdata import TcrData
from os import listdir
from os.path import isfile, join
from math import log


datadir = "../data/repTCRb/"
pepdir = "../data/peptideTCRab/"
figdir = "../results/timescatter/"
respfile = "../data/groups.csv"

# Get the responder type data

responders = pd.read_csv(respfile,sep=",")
responders = responders.set_index('Vaccinee')

# Get all rep files at time 0 from directory
files = [f for f in listdir(datadir) if (isfile(join(datadir, f)) and "B0.tsv" in f)]


results_list = []


for f in files:
    root = f[0:-7]
    print(root)
    pre = TcrData()
    pre.read_immunoseq(join(datadir, root + '_B0.tsv'))
    post = TcrData()
    post.read_immunoseq(join(datadir, root + '_B60.tsv'))

    # print(pre.df)

    # Calculate overlap

    overlaptcr = list(pre.df.index[np.in1d(pre.df.index, post.df.index, assume_unique=True)])

    print('B0-B60 overlap: ' + str(len(overlaptcr)))

    print(pre.df.loc[overlaptcr]['freq'])

    # Get peptide specific TCRs

    pepfiles = [p for p in listdir(pepdir) if (isfile(join(pepdir, p))
                                               and (root + '_' in p)
                                               and ('beta.txt' in p)
                                               and not ('VZV' in p))]

    pepdata = dict()
    peptotdatatmp = dict()

    for p in pepfiles:

        print(p)

        m = re.match("H\d+_(P+\d*)_([ab])", p)

        if m:

            peptide = m.group(1)
            sample = m.group(2)

            if sample not in peptotdatatmp:
                peptotdatatmp[sample] = dict()

            pep = TcrData()
            pep.read_mixcr(join(pepdir, p))

            if peptide not in peptotdatatmp[sample]:
                peptotdatatmp[sample][peptide] = set()
            peptotdatatmp[sample][peptide].update(set(pep.raw['cdr3'].unique()))

    peptotdata = dict()
    for peptide in peptotdatatmp['a']:
        print(peptide)
        #Only keep those TCRs that occur in both in vitro assays
        peptotdata[peptide] = peptotdatatmp['a'][peptide].intersection(peptotdatatmp['b'][peptide])

        #Calculate overlap between assay TCRs and overlaptcr (those that occur at day 0 and day 60)
        pepoverlap = peptotdata[peptide].intersection(set(overlaptcr))

        if(len(pepoverlap) > 0):
            if peptide not in pepdata:
                pepdata[peptide] = set()
            pepdata[peptide].update(pepoverlap)
            print(pepdata[peptide])
            print(len(pepdata[peptide]))


    # Calculate some statistics

    resdict = {}

    resdict['volunteer'] = root
    resdict['B0total'] = len(pre.df.index)
    resdict['B0entropy'] = pre.entropy()
    resdict['B60total'] = len(post.df.index)
    resdict['B60entropy'] = post.entropy()
    resdict['B0B60overlap'] = len(set(pre.df.index).intersection(set(post.df.index)))
    resdict['B0reads'] = pre.df['count'].sum()
    resdict['B60reads'] = post.df['count'].sum()
    resdict['responder'] = responders.loc[root]['Status_2']
    if 'PP' in peptotdata:
        resdict['PPTotal'] = len(peptotdata['PP'])
        if 'PP' in pepdata:
            resdict['PPIntersect'] = len(pepdata['PP'])
            resdict['PPIntInc'] = len([p for p in pepdata['PP']
                                       if pre.df.loc[p]['freq']
                                       < post.df.loc[p]['freq']])
            resdict['PPIntB0freq'] = sum([pre.df.loc[p]['freq'] for p in pepdata['PP']])
            resdict['PPIntB60freq'] = sum([post.df.loc[p]['freq'] for p in pepdata['PP']])
        else:
            resdict['PPIntersect'] = 0
            resdict['PPIntInc'] = 0
            resdict['PPIntB0freq'] = 0
            resdict['PPIntB60freq'] = 0
        resdict['PPB0'] = len(peptotdata['PP'].intersection(pre.df.index))
        resdict['PPB60'] = len(peptotdata['PP'].intersection(post.df.index))

    results_list.append(resdict)

    # Plot

    plt.scatter(pre.df.loc[overlaptcr]['freq'].apply(log),
                post.df.loc[overlaptcr]['freq'].apply(log),
                label="Aspecific")
    plt.xlabel(root + ' Time 0')
    plt.ylabel(root + ' Time 60')
    ax = plt.gca()
    low_x, high_x = ax.get_xlim()
    low_y, high_y = ax.get_ylim()
    plt.plot([max(low_x, low_y), min(high_x, high_y)], [max(low_x, low_y), min(high_x, high_y)], 'k--')

    colorcode = {'PP': '#FF0000'}
    for i in range(1, 60):
        colorcode['P'+str(i)] = '#' + str(int(100-i)) + str(int(30 + (i**2) % 70)) + '0' + str(int(i % 10))

    for peptide in pepdata:
        if peptide == 'PP':
            plt.scatter(pre.df.loc[pepdata[peptide]]['freq'].apply(log),
                        post.df.loc[pepdata[peptide]]['freq'].apply(log),
                        c=colorcode[peptide], label=peptide)

    lgd = plt.legend(loc='center left', bbox_to_anchor=(1, 0.5))
    plt.savefig(figdir + root + '_PP.pdf', bbox_extra_artists=(lgd,), bbox_inches='tight')
    plt.close()

    #quit()

resultsdf = pd.DataFrame(results_list)
resultsdf.to_csv(figdir + 'volunteercounts.txt')
